pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1947,8 +1933,34 @@ protected void triggerFlush() {
         }
     }
 
+    // must acquire semaphore before calling
+    private void maybeTriggerBatchTimerTask() {
+        if (this.batchTimerTask == null) {
+            this.batchTimerTask = cnx().ctx().executor()
+                    .schedule(catchingAndLoggingThrowables(() -> {
```

Comment:
This could be refactored further. A separate method would be helpful to reduce the indention levels.

Reply:
I agree.

Reply:
Also returning early could help: 


```java
   if (this.batchTimerTask != null) {
       return;
   }
```